lib/EarlyFunctionAnalysis/AttachDebugInfo.cpp

```diff
@@ -0,0 +1,137 @@
+
+    for (auto &BB : F) {
+      CurrentDebugLocation = nullptr;
+
+      if (!GCBI.isJumpTarget(&BB)) {
+        revng_log(Log, " Not a jump target " << BB.getName());
+        continue;
+      }
+
+      const efa::BasicBlock *Block = FM.findBlock(GCBI, &BB);
+      if (!Block) {
+        revng_log(Log, " No metadata for " << BB.getName());
+        continue;
+      }
+
+      revng_log(Log, " " << BB.getName() << ": " << Block->Start.toString());
+      for (auto &I : BB) {
+        if (auto *Call = getCallTo(&I, "newpc")) {
+          MetaAddress NewPC = GeneratedCodeBasicInfo::getPCFromNewPC(Call);
+          // We cloned `line:` only, but we need full MetaAddress that we keep
+          // as `DISubprogram`.
+          auto SPFlags = DISubprogram::toSPFlags(false, /* isLocalToUnit */
+                                                 true, /* isDefinition*/
+                                                 false /* isOptimized */);
+          auto Type = DIB.createSubroutineType(DIB.getOrCreateTypeArray(None));
+
+          // Let's make the debug location that points back to the binary.
+          std::string NewDebugLocation = "/instruction/" + FM.Entry.toString()
+                                         + "/" + Block->Start.toString() + "/"
+                                         + NewPC.toString();
+          auto Subprogram = DIB.createFunction(File, // Scope
+                                               NewDebugLocation, // Name
+                                               StringRef(), // LinkageName
+                                               File, // File
+                                               1, // LineNo
+                                               Type, // Ty (subroutine type)
+                                               1, // ScopeLine
+                                               DINode::FlagPrototyped, // Flags
+                                               SPFlags);
+          DIB.finalizeSubprogram(Subprogram);
+          auto InlineLocationForMetaAddress = DILocation::get(M.getContext(),
+                                                              NewPC.address(),
+                                                              0,
+                                                              RootSubprogram,
+                                                              nullptr);
+          // Represent debug info for all the isolated functions as if they were
+          // inlined in the root.
+          auto InlineLoc = DILocation::get(M.getContext(),
+                                           NewPC.address(),
+                                           0,
+                                           Subprogram,
+                                           InlineLocationForMetaAddress);
+          CurrentDebugLocation = InlineLoc;
+          I.setDebugLoc(CurrentDebugLocation);
+        } else {
+          I.setDebugLoc(CurrentDebugLocation);
+        }
```

Comment:
Consider outlining in a `handleBasicBlock`. Lines are very long and outlining you can gain ~6 chars `:)`